signer/src/network/libp2p/swarm.rs

```diff
@@ -70,22 +85,66 @@ impl SignerBehavior {
-            .add_seed_addresses(config.seed_addresses);
-
+            .add_seed_addresses(&config.seed_addresses);
         let bootstrap = bootstrap::Behavior::new(bootstrap_config);
 
         Ok(Self {
             gossipsub: Self::gossipsub(&keypair)?,
             mdns,
-            kademlia: Self::kademlia(local_peer_id),
+            kademlia,
             ping: Default::default(),
             identify,
             autonat_client,
             autonat_server,
             bootstrap,
+            connection_limits: Self::connection_limits(&config),
         })
     }
 
+    fn connection_limits(config: &SignerSwarmConfig) -> connection_limits::Behaviour {
+        // The number of signers is the number of signers in the set, minus one
+        // for the local signer. This is used to calculate the connection limits.
+        // Note: We use `max(2)` here to ensure that we always have limits
+        // that support at least two signers.
+        let num_signers = config.num_signers.saturating_sub(1).max(2) as u32;
+
+        // Allow `num_signers` * 3 connections to be established at once. This
+        // allows room for one incoming/outgoing connection to each signer, plus
+        // one additional connection per signer as a buffer for protocols such
+        // as autonat.
+        let max_established = num_signers.saturating_mul(3);
+
+        // Allow `num_signers` * 2 _incoming_ connections to be established at
+        // once. This allows room for one incoming connection from each signer,
+        // plus one additional connection per signer as a buffer for protocols
+        // such as autonat.
+        let max_established_incoming = num_signers.saturating_mul(2);
+
+        // Allow for one incoming, one outgoing and one additional connection per
+        // peer.
+        let max_established_per_peer = 3;
+
+        // Allow for `num_signers` _incoming_ connections to be pending at once.
+        // This is conservative, but realistic as we wouldn't expect each signer
+        // to be attempting to open multiple connections to us simultaneously,
+        // even during bootstrapping.
+        let max_pending_incoming = num_signers;
+
+        // Allow for `num_signers` _outgoing_ connections to be pending at once.
+        // This is conservative, but realistic as we wouldn't expect to be dialing
+        // more than `num_signers` signers at once, even during bootstrapping.
+        let max_pending_outgoing = num_signers;
+
+        let limits = connection_limits::ConnectionLimits::default()
+            .with_max_established(Some(max_established))
+            .with_max_established_incoming(Some(max_established_incoming))
+            .with_max_established_per_peer(Some(max_established_per_peer))
+            .with_max_pending_incoming(Some(max_pending_incoming))
+            .with_max_pending_outgoing(Some(max_pending_outgoing));
+
+        connection_limits::Behaviour::new(limits)
+    }
+
```

Comment:
To reviewers: please put extra focus on this block, we need this calculations to be sane with a volatile signing set, but note that these limits are currently still counting on the fact that the signing set is a closed network.

Reply:
Oh we do do this already in `main.rs`, though I just updated it to use the `SignerState` view instead of the config specifically (as that should be more future-proof) in [00ebd4d](https://github.com/stacks-sbtc/sbtc/pull/1568/commits/00ebd4d55ed8acf882456efc9f770241abf0a4c0), but it is capped at 128 via `MAX_KEYS`.